srcs/Server.cpp

```diff
@@ -2,117 +2,149 @@
+		for (int i = 0; i < fd_count; i++)
+		{
+			// Check if descriptor has data available
+			if (pfds[i].revents & POLLIN)
+			{
+				if (pfds[i].fd == listenSocket.getSocketFD())
+					handleConnection();
+				else
+					handleRequest(i);
+			}
 		}
-		break ;
 	}
-	// When the loop is complete, we should have a socket bound to our address, we can free now
-	freeaddrinfo(servinfo);
-	if (p == NULL)
-		throw Socket::SocketBindException();
-	listenSocket.listenOnSocket();
-
-	sa.sa_handler = sigchld_handler;
-	sigemptyset(&sa.sa_mask);
-	sa.sa_flags = SA_RESTART;
-	if (sigaction(SIGCHLD, &sa, NULL) < 0)
+}
+
+// PRIVATE METHODS
+void	Server::handleRequest(int i) {
+	int	nbytes;
+	int sender_fd;
+	char buf[1024];    // Buffer for client data
+
+	sender_fd = pfds[i].fd;
+	memset(buf, 0, 1024);
+	nbytes = recv(sender_fd, buf, sizeof(buf), 0);
+	if (nbytes <= 0)
 	{
-		perror("sigaction");
-		exit(1);
+		if (nbytes == 0)
+			std::cout << "server: Socket " << sender_fd << " hung up" << std::endl;
+		else if (nbytes < 0)
+			std::cout << "server: recv error" << std::endl;
+		close(sender_fd);
+		dropConnection(i);
 	}
-	std::cout << "server: waiting for connections..." << std::endl;
-	while (1)
+	else
+	{
+		// we got some data hurrah!
+		// TODO: parsing -> put inside of a request class
+		// But for now, let's just print what we received
+		std::cout << "Received: " << buf << std::endl;
+	}
+}
+
+void	Server::addConnection(int newfd) {
+	// This function adds a new file descriptor to the set of pfds
+	// if we don't have room, we need to realloc
+	if (fd_count == fd_size)
 	{
-		try {
-			currSocket.acceptConnection(listenSocket.getSocketFD());
-		} catch (std::exception& e){
-			continue ; // If connection accept fails, just try again later
+		fd_size *= 2;
+		struct pollfd *pfds_new = new pollfd[fd_size];
+		for (int i = 0; i < fd_count; i++)
+		{
+			pfds_new[i].fd = pfds[i].fd;
+			pfds_new[i].events = pfds[i].events;
+			pfds_new[i].revents = pfds[i].revents;
 		}
-		currSocket.printConnection();
-		// Not sure how large to make the recv buffer... Explanation from link below
-		// https://stackoverflow.com/questions/2862071/how-large-should-my-recv-buffer-be-when-calling-recv-in-the-socket-library
-		memset(recv_buff, 0, BUFF_SIZE);
-		recv(currSocket.getSocketFD(), recv_buff, BUFF_SIZE, 0);
-		time_t timetoday;
-		time(&timetoday);
-		std::cout << "TIMESTAMP: " << std::asctime(localtime(&timetoday)) << std::endl;
-		std::cout << "Received: " << recv_buff << std::endl;
-		// Server response should be here:
-		// server.sendResponse();
-		// Fork and send HTTP/1.1 200 OK\r\n ... etc
-		close(currSocket.getSocketFD());
+		delete[] pfds;
+		pfds = pfds_new;
 	}
-	std::cout << "Hurray we made it!" << std::endl;
-}
\ No newline at end of file
+	pfds[fd_count].fd = newfd;
+	pfds[fd_count].events = POLLIN;
+	fd_count++;
+}
+
+void	Server::dropConnection(int i) {
+	// copy the last over this one.
+	pfds[i] = pfds[fd_count - 1];
```

Comment:
I am not understanding this function - as far as I understand, this line is replacing the element indicated (`pfds[i]`) with the previous `i - 1`? Could you please explain me better this mechanism? :)

Reply:
In the network programming book, this is suggested as the most straightforward way to drop a connection from the list of file descriptors. 

Basically, if we want to drop connection `i`, we copy over the last file descriptor in the list of `*pfds` and then reduce the current count of tracked fds by 1. If we would open a new connection after this, it would just overwrite the content of what was once there without any issue.

Reply:
Clear :)